src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -166,11 +166,139 @@ public ReadOnlySequence<T> Slice(long start, long length)
+                            new SequencePosition(beginObject, beginIndex + (int)length) :
+                            SeekMultiSegment(beginSegment.Next, endObject, endIndex, length - currentLength);
+                    }
+                    else
+                    {
+                        if (endIndex - beginIndex < length)
+                            ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+                        // beginIndex + offset <= int.MaxValue
+                        Debug.Assert(length <= int.MaxValue - beginIndex);
+
+                        end = new SequencePosition(beginObject, beginIndex + (int)length);
+                    }
+                }
+            }
+            else
+            {
+                if (endIndex - startIndex < start)
+                    ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+                // startIndex + offset <= int.MaxValue
+                Debug.Assert(start <= int.MaxValue - startIndex);
+
+                startIndex += (int)start;
+                begin = new SequencePosition(startObject, startIndex);
+
+                if (endIndex - startIndex < length)
+                    ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+                // beginIndex + offset <= int.MaxValue
+                Debug.Assert(length <= int.MaxValue - startIndex);
+
+                end = new SequencePosition(startObject, startIndex + (int)length);
+            }
+
             return SliceImpl(begin, end);
         }
 
+        private static bool InRange(uint value, uint start, uint end)
+        {
+            // _sequenceStart and _sequenceEnd must be well-formed
+            Debug.Assert(start <= int.MaxValue);
+            Debug.Assert(end <= int.MaxValue);
+            Debug.Assert(start <= end);
+
+            // The case, value > int.MaxValue, is invalid, and hence it shouldn't be in the range.
+            // If value > int.MaxValue, it is invariably greater than both 'start' and 'end'.
+            // In that case, the experession simplifies to value <= end, which will return false.
+
+            // The case, value < start, is invalid.
+            // In that case, (value - start) would underflow becoming larger than int.MaxValue.
+            // (end - start) can never underflow and hence must be within 0 and int.MaxValue. 
+            // So, we will correctly return false.
+
+            // The case, value > end, is invalid.
+            // In that case, the expression simplifies to value <= end, which will return false.
+            // This is because end > start & value > end implies value > start as well.
+
+            // In all other cases, value is valid, and we return true.
+
+            // Equivalent to: return (start <= value && value <= start)
+            return (value - start) <= (end - start);
+        }
+
+        private static bool InRange(ulong value, ulong start, ulong end)
```

Comment:
No, we can't do that since end can be negative. We are relying on the cast from long to ulong to catch overflow cases (like negative end).

Reply:
> No, we can't do that since end can be negative. We are relying on the cast from long to ulong to catch overflow cases (like negative end).

Are you correct? https://dotnetfiddle.net/kJjenR

Reply:
My point is, we can't assert `end >= 0` if end is of type long. But, yes, we could cast it to ulong inside. Let me try.

Reply:
I think I will keep it as is, especially since we may change GetIndex to return uint instead. It is mainly a cosmetic change atm and we can always do it in the future.